contract/contracts/forward_name_resolver/forward_name_resolver.ral

```diff
@@ -0,0 +1,259 @@
+            nameNftIndex[name] = totalSupply
+        } else {
+            nameNftIndex.insert!(
+                caller,
+                name,
+                totalSupply
+            )
+        }
+
+        handleRewardTokens{caller -> rewardTokenId: tokenRemaining!(caller, rewardTokenId)}(
+            caller,
+            name
+        )
+
+        return createName{caller -> ALPH: minimalContractDeposit!()}(
+            caller,
+            totalSupply,
+            name,
+            capitalisation
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn setAddress(name: ByteVec, newAddress: Address) -> () {
+        let caller = callerAddress!()
+        let nft = getNftByName(name)
+        let tokenId = tokenId!(nft)
+
+        checkCaller!(tokenRemaining!(caller, tokenId) == 1, ErrorCodes.OnlyNftOwnerAllowed)
+        nft.setAddress(newAddress)
+
+        emit AddressSet(
+            nft.getNFTIndex(),
+            name,
+            newAddress
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn setCapitalisation(name: ByteVec, newCapitalisation: ByteVec) -> () {
+        let caller = callerAddress!()
+        let nft = getNftByName(name)
+        let tokenId = tokenId!(nft)
+        let isNftHolder = tokenRemaining!(caller, tokenId) == 1
+        let isCurrentAddress = caller == nft.getAddress()
+
+        checkCaller!(isNftHolder || isCurrentAddress, ErrorCodes.OnlyNftOwnerOrHolderAllowed)
+        nft.setCapitalisation(newCapitalisation)
+
+        emit CapitalisationSet(
+            nft.getNFTIndex(),
+            name,
+            newCapitalisation
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn renewName(name: ByteVec) -> () {
+        let caller = callerAddress!()
+        let nft = getNftByName(name)
+        let tokenId = tokenId!(nft)
+        checkCaller!(tokenRemaining!(caller, tokenId) == 1, ErrorCodes.OnlyNftOwnerAllowed)
+
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+
+        assert!(canRenew(nft), ErrorCodes.CannotRenewName)
+
+        handleRewardTokens{caller -> rewardTokenId: tokenRemaining!(caller, rewardTokenId)}(
+            caller,
+            name
+        )
+
+        let now = blockTimeStamp!()
+        let expires = now + renewLength
+        nft.setExpires(expires)
+
+        emit NameRenewed(
+            nft.getNFTIndex(),
+            name,
+            caller,
+            expires
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn deleteExpired(name: ByteVec) -> () {
+        let caller = callerAddress!()
+        let existingNft = getNftByName(name)
+        let tokenId = tokenId!(existingNft)
+        checkCaller!(tokenRemaining!(caller, tokenId) == 1, ErrorCodes.OnlyNftOwnerAllowed)
+
+        let nftIndex = existingNft.getNFTIndex()
+        let expired = hasExpired(existingNft)
+        assert!(expired, ErrorCodes.NameHasNotExpired)
+        existingNft.delete(caller)
```

Comment:
Does it make sense to burn the token here?

Reply:
Yes thats a nice improvement

Reply:
Actually I want to make this function so that anybody can clean up expired NFTs

Reply:
I am guessing someone can set up a bot to monitor the names that are soon expired and delete them since it could be profitable, while previous owner might just have forgotten to renew. ENS introduced a grace period of 90 days which could mitigate this issue, not sure it is worth adding here.

Reply:
There is a 30 day grace period before it expires where only the NFT holder can renew their Name. We don't mind if they can't get their deposit back